### ISSUE TITLE
Publish ROS package actions

### ISSUE THREAD

**Issue by patrick-5546** [UBCSailbot/network_systems#11]
### Purpose

Each action that is used across our ROS packages should be published so that updates can be pushed to all ROS package repositories automatically.

### Changes

- Create repository for lint action
- Create repository for test action
- Use published actions in this repository

### Final To do
- [ ] Merge sailbot_workspace PR without switching branches
- [ ] Manually delete branch and re-check "Automatically delete head branches"
- [ ] Point network_systems PR to sailbot_workspace main branch, updated required jobs, then merge
- [ ] Create sailbot_workspace PR to point to main branch

### Resources

- https://docs.github.com/en/actions/creating-actions

**Comment by colinli02** [comment]
Status: 
Awaiting repo made for actions (lack of permissions)

**Comment by patrick-5546** [comment]
Explore https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow

**Comment by colinli02** [comment]
> Explore https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow
> 
> * Reuse network_systems CI in other repos
> * Possible to reuse sailbot_workspace CI in network_systems?

Should there be 2 files in total or 4?
e.g.
- run network_systems CI test actions
- run network_systems CI lint actions
- run sailbot_workspace CI test actions
- run sailbot_workspace CI lint actions 
OR
- run network & sailbot lint actions
- run network & sailbot test actions

I have the code on my local machine but I'll want to know which one it is before a push, thank you

**Comment by colinli02** [comment]
My checklist:
- [ ] Network systems & sailbot actions are accessible by others [(see](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#allowing-access-to-components-in-a-private-repository)
- [ ] Create repository for actions? (changed plans, to re-use workflows instead)
- [ ] Use published actions in this repository
- [ ] Make .yml workflow to use network_systems CI
- [ ] Make .yml workflow to use sailbot_workspace CI 

Additional Notes:
- to edit

**Comment by colinli02** [comment]
> I want to have as little duplicate code in workflows as possible. Ideally it would look like this: both sailbot_workspace and subteam repositories use workflows defined in sailbot_workspace. The subteam repositories would call it with different arguments to only run the tests that are applicable to their codebase and only run those tests on their codebase. With what you know now, do you think this is possible?

The way our current workflows in sailbot_workspace are written involves using our local actions .yml within .github/actions.

```
  colcon-test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v3
 ```

This means if we call our workflows from within the same repository, it's super easy, but that's not what we want here, since we want to call it from another repository (e.g. network systems), and have minimal duplicate code, which means somehow just calling sailbot_workspace workflows and not having to duplicate the action files into our .github folder

```
call-workflow-1-in-workflows: # Calls tests.yaml, requires to duplicate code into own workflow, so duplicate code needed
    #uses: actions/checkout@v3
    uses: UBCSailbot/sailbot_workspace/.github/workflows/tests.yaml@user/colinli02/11-Publish-ROS-package-actions
```

So if we want to call sailbot_workspace's workflow, we would want to be able to somehow checkout the actions into our sub-team repository. However that means we can't straight up call "tests.yaml" from sailbot_workspace, since we need to somehow checkout the ./.github/actions
# All solutions:
## A) 
https://github.com/orgs/community/discussions/26766 Manually copy the entire .github file over into our sub-team repos, but this defeats the purpose since we want to minimize code duplication without copying sailbot_workspace's workflow over and having sailbot_workspace as a "central" workflow.
## B)
 Henry has a smart workaround where he "clone[s] sailbot_workspace:main so that sailbot_workspace is the root, and then clone network_systems into it and run the actions defined in sailbot_workspace". for example https://github.com/UBCSailbot/network_systems/blob/main/.github/workflows/lint.yaml This is probably our best solution, to follow Henry's method since we can still have our central workflows in sailbot_workspace, however comes at the cost of still not having a completely centralized workflow in sailbot_workspace and just calling it, since there are chunks of code where it calls another public repo's workflow

```
  # https://github.com/gaurav-nelson/github-action-markdown-link-check
  markdown-link-check:
    runs-on: ubuntu-latest
    steps:
    - name: Check out code
      uses: actions/checkout@v3

    - name: Run linter
      uses: gaurav-nelson/github-action-markdown-link-check@v1
      with:
        config-file: .markdown-link-check.json
        folder-path: .
```

 which means if we were to want to change these parts, we would have to change ALL the sub-teams individually.
This also means we have to make minimal modifications to parts such as:
 - name: Checkout code
        uses: actions/checkout@v3
        with:
          path: src/network_systems # change here to your repo src name 
     
## C)
 If we were to want a completely centralized workflow in sailbot_workspace and call it, we would have to move these "public" repos ex:
 ```
 - name: Run linter
      uses: gaurav-nelson/github-action-markdown-link-check@v1
      with:
        config-file: .markdown-link-check.json
        folder-path: .
  ```
into our sailbot_workspace into actions and call it, but from what I have so far looked into it I don't think this is possible.


**TLDR: Unfortunately, from what I have researched the ability to call sailbot_workspace's workflows directly without any code duplication is not possible at this moment. (edited)** 





**So my conclusion is if we wanted to only use sailbot_workspace's workflows, we have to copy Henry's current network_systems CI implementation methods, however unfortunately that means we don't have a way of completely having a central workflow in sailbot_workspace to "just call" since there are parts that we cannot keep in sailbot_workspace only.**

**Comment by patrick-5546** [comment]
Thanks for your extensive research into this matter, this is definitely not a simple problem and I'm glad to see you rise to the occasion.

Responding to the following comment:

> I have taken a long look into this and unfortunately it's not as simple as just calling the re-useable workflow even with on.workflow_call implemented onto the .yaml. This is because these re-useable workflows do not require checkouts of actions like we have set-up.

### Background

#### Current sailbot Workspace CI
```yml
  colcon-test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Test
        uses: ./.github/actions/test/
```

#### Current network system CI
```yml
  colcon-test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout workspace
        uses: actions/checkout@v3
        with:
          repository: UBCSailbot/sailbot_workspace

      - name: Checkout code
        uses: actions/checkout@v3
        with:
          path: src/network_systems

      - name: Test
        uses: ./.github/actions/test/
        env:
          DISABLE_VCS: true
```

#### Goal
Update sailbot workspace CI so that it can run like current sailbot workspace CI when called by sailbot_workspace and like current network systems CI when called by network_systems.

### Proposed solution

I believe that this behavior can be achieved by using [`jobs.<job_id>.steps[*].if`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif) and [`on.workflow_call.inputs.<input_id>.type`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype).

#### Updated sailbot workspace CI
```yml
  colcon-test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout workspace
        uses: actions/checkout@v3
        with:
          repository: ${{ inputs.repository }}

      - name: Checkout code
        if: ${{ inputs.repository }} != ${{ github.repository }}
        uses: actions/checkout@v3
        with:
          path: src/${{ inputs.repository }}

      - name: Test
        uses: ./.github/actions/test/
        env:
          DISABLE_VCS: ${{ inputs.disable-vcs }}
```

Would something like this work?

### Resources
- https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow

**Comment by colinli02** [comment]
Alright, here's an update on my current progress:

# Calling Sailbot_workspace CI within network_systems:


## Calling test action only

**Status:** ✔️

Network_systems code:
```
# Calls sailbot ws CI within network systems
jobs:
# Calls our workflow in sailbot_workspace 
# UBCSailbot/sailbot_workspace/tree/main/.github/workflows
  call-workflow-1-in-sailbot:
        uses: UBCSailbot/sailbot_workspace/.github/workflows/tests.yml@user/colinli02/11-Publish-ROS-package-actions
        with:
            config-path: .github/tests.yml
        secrets: inherit

```
CI Code
```
name: Tests

on:
  workflow_call:

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

on:
  workflow_call:
    inputs:
      disable-vcs:
        required: true
        type: string
      repository:
        required: true
        type: string

jobs:
  colcon-test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout workspace
        uses: actions/checkout@v3
        with:
          repository: UBCSailbot/sailbot_workspace

      - name: Checkout code
        if: ${{ inputs.repository }} != ${{ github.repository }}
        uses: actions/checkout@v3
        with:
          path: src/${{ inputs.repository }}

      - name: Test
        uses: ./.github/actions/test/
        env:
          DISABLE_VCS: ${{ inputs.disable-vcs }}

```
... to edit

**Comment by colinli02** [comment]
# Useful references for the future:
https://github.com/actions/checkout
https://github.com/actions/checkout/issues/24

**Comment by colinli02** [comment]
# Issue with PR:

Now that I have squished its commit history, I can't create a PR, nor can I ```git pull``` because of its different histories.

However based on the above information, I think its easy to just manually add it in, test it to make sure it works, then commit it from the main branch. Unless there is a solution to this issue.


The only required change in code to verify it works:

```
        uses: UBCSailbot/sailbot_workspace/.github/workflows/tests.yml@user/colinli02/11-Publish-ROS-package-actions
```
into
```
        uses: UBCSailbot/sailbot_workspace/.github/workflows/tests.yml@umain 
```

**Comment by patrick-5546** [comment]
Just create a new branch and copy over the changes you made

**Comment by patrick-5546** [comment]
Can `git reset` to specific commit ID
